File: threedollar-api-core/src/main/java/com/depromeet/threedollar/application/config/redis/EmbeddedRedisConfig.java
package com.depromeet.threedollar.application.config.redis;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;
import org.springframework.data.redis.connection.RedisConnectionFactory;
import org.springframework.data.redis.connection.lettuce.LettuceConnectionFactory;
import redis.embedded.RedisServer;

import javax.annotation.PostConstruct;
import javax.annotation.PreDestroy;
import java.io.IOException;

import static com.depromeet.threedollar.common.utils.ProcessUtils.findAvailableRandomPort;
import static com.depromeet.threedollar.common.utils.ProcessUtils.isRunningPort;

/**
 * 로컬 및 테스트용 임베디드 Redis Server 설정.
 */
@Slf4j
@RequiredArgsConstructor
@Profile("local")
@Configuration
public class EmbeddedRedisConfig {

    private final RedisProperties properties;

    private RedisServer redisServer;

    @Value("${spring.redis.port}")
    private int port;

    @PostConstruct
    public void startRedis() throws IOException {
        if (redisServer == null || !redisServer.isActive()) {
            port = isRunningPort(port) ? findAvailableRandomPort() : port;
            redisServer = RedisServer.builder()
                .port(port)
                .setting("maxmemory 128M")
                .build();
            redisServer.start();
            log.info("임베디드 레디스 서버가 기동되었습니다. port: {}", port);
        }
    }

    @PreDestroy
    public void stopRedis() {
        if (redisServer != null) {
            redisServer.stop();
            log.info("임베디드 레디스 서버가 종료됩니다");
        }
    }

    @Bean
    public RedisConnectionFactory embeddedRedisConnectionFactory() {
        return new LettuceConnectionFactory(properties.getHost(), port);
    }

}

File: threedollar-boss-api/src/main/kotlin/com/depromeet/threedollar/foodtruck/api/config/redis/EmbeddedRedisConfig.kt
package com.depromeet.threedollar.foodtruck.api.config.redis

import com.depromeet.threedollar.common.utils.ProcessUtils
import lombok.RequiredArgsConstructor
import lombok.extern.slf4j.Slf4j
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Profile
import org.springframework.data.redis.connection.RedisConnectionFactory
import org.springframework.data.redis.connection.lettuce.LettuceConnectionFactory
import redis.embedded.RedisServer
import javax.annotation.PostConstruct
import javax.annotation.PreDestroy

/**
 * 로컬 및 테스트용 임베디드 Redis Server 설정.
 */
@Slf4j
@RequiredArgsConstructor
@Profile("local")
@Configuration
class EmbeddedRedisConfig(
    private val properties: RedisProperties
) {

    private lateinit var redisServer: RedisServer
    private var port: Int = 0

    @PostConstruct
    fun startRedis() {
        port = if (ProcessUtils.isRunningPort(properties.port)) ProcessUtils.findAvailableRandomPort() else properties.port
        redisServer = RedisServer(port)
        redisServer.start()
    }

    @PreDestroy
    fun stopRedis() {
        redisServer.stop()
    }

    @Bean
    fun embeddedRedisConnectionFactory(): RedisConnectionFactory {
        return LettuceConnectionFactory(properties.host, port)
    }

}
